File: src/babylon/engine/event_bus.py
"""Event system for decoupled communication in the simulation.

This module provides a publish/subscribe event bus that enables loose coupling
between simulation components. Events are immutable data objects that carry
information about state changes.

Sprint 3: Central Committee (Dependency Injection)
Epoch 1→2 Bridge: Added EventInterceptor pattern for adversarial mechanics.
"""

from __future__ import annotations

import logging
from collections import defaultdict
from collections.abc import Callable
from dataclasses import dataclass, field
from datetime import datetime
from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from babylon.engine.interceptor import (
        BlockedEvent,
        EventInterceptor,
        WorldContext,
    )

logger = logging.getLogger(__name__)


@dataclass(frozen=True)
class Event:
    """Immutable event representing a simulation occurrence.

    Events are frozen dataclasses to ensure they cannot be modified
    after creation, maintaining integrity of the event history.

    Attributes:
        type: Event type identifier (e.g., "tick", "rupture", "synthesis")
        tick: Simulation tick when the event occurred
        payload: Event-specific data dictionary
        timestamp: Wall-clock time when event was created
    """

    type: str
    tick: int
    payload: dict[str, Any]
    timestamp: datetime = field(default_factory=datetime.now)


# Type alias for event handlers
EventHandler = Callable[[Event], None]


class EventBus:
    """Publish/subscribe event bus for simulation components.

    The EventBus enables decoupled communication between systems.
    Components can subscribe to specific event types and will be
    notified when events of that type are published.

    All published events are stored in history for replay/debugging.

    Epoch 1→2 Bridge: Supports optional interceptor chain for adversarial
    mechanics. If no interceptors are registered, events flow through with
    zero overhead (backwards compatible).

    The interceptor chain processes events before emission:
    - Interceptors are sorted by priority (higher runs first)
    - Each interceptor can ALLOW, BLOCK, or MODIFY the event
    - If blocked, the event is logged and not emitted
    - If modified, the modified event continues through the chain

    Example:
        >>> bus = EventBus()
        >>> def on_tick(event: Event) -> None:
        ...     print(f"Tick {event.tick}: {event.payload}")
        >>> bus.subscribe("tick", on_tick)
        >>> bus.publish(Event(type="tick", tick=1, payload={"value": 42}))
        Tick 1: {'value': 42}
    """

    def __init__(self) -> None:
        """Initialize an empty event bus."""
        self._subscribers: dict[str, list[EventHandler]] = defaultdict(list)
        self._history: list[Event] = []
        self._interceptors: list[EventInterceptor] = []
        self._blocked_events: list[BlockedEvent] = []

    def subscribe(self, event_type: str, handler: EventHandler) -> None:
        """Subscribe a handler to receive events of a specific type.

        Args:
            event_type: The type of events to subscribe to
            handler: Callable that receives Event objects
        """
        self._subscribers[event_type].append(handler)

    def register_interceptor(self, interceptor: EventInterceptor) -> None:
        """Register an interceptor to process events before emission.

        Interceptors are sorted by priority (higher first) each time
        an event is published. Multiple interceptors with the same
        priority execute in registration order.

        Args:
            interceptor: The interceptor to register.

        Example:
            >>> from babylon.engine.interceptor import EventInterceptor
            >>> bus = EventBus()
            >>> bus.register_interceptor(my_security_interceptor)
        """
        self._interceptors.append(interceptor)

    def unregister_interceptor(self, interceptor: EventInterceptor) -> None:
        """Remove an interceptor from the chain.

        Args:
            interceptor: The interceptor to remove.

        Raises:
            ValueError: If the interceptor is not registered.
        """
        self._interceptors.remove(interceptor)

    def publish(self, event: Event, context: WorldContext | None = None) -> None:
        """Publish an event to all subscribed handlers.

        If interceptors are registered, the event passes through the
        interceptor chain first. If any interceptor blocks the event,
        it is logged to the blocked events audit channel and not emitted.

        The event is stored in history only if it passes all interceptors.

        Args:
            event: The event to publish.
            context: Optional world context for interceptors.
                    Required for Epoch 2 adversarial mechanics.
        """
        # Fast path: no interceptors = zero overhead for Epoch 1
        if not self._interceptors:
            self._history.append(event)
            self._emit_to_handlers(event)
            return

        # Slow path: run through interceptor chain
        processed_event = self._process_interceptors(event, context)

        if processed_event is not None:
            # Event passed all interceptors - emit to subscribers
            self._history.append(processed_event)
            self._emit_to_handlers(processed_event)

    def _process_interceptors(self, event: Event, context: WorldContext | None) -> Event | None:
        """Process event through the interceptor chain.

        Args:
            event: The event to process.
            context: Optional world context for interceptors.

        Returns:
            The (possibly modified) event if allowed, None if blocked.
        """
        # Import here to avoid circular import at module level
        from babylon.engine.interceptor import BlockedEvent

        current_event: Event = event

        # Sort by priority (higher first), stable sort preserves registration order
        sorted_interceptors = sorted(
            self._interceptors,
            key=lambda i: i.priority,
            reverse=True,
        )

        for interceptor in sorted_interceptors:
            result = interceptor.intercept(current_event, context)

            if result.is_blocked:
                # BLOCKED - log to audit channel and stop
                blocked_record = BlockedEvent(
                    event=event,  # Log original event for auditability
                    interceptor_name=interceptor.name,
                    reason=result.reason,
                )
                self._blocked_events.append(blocked_record)

                logger.warning(
                    "Event %s BLOCKED by %s: %s",
                    event.type,
                    interceptor.name,
                    result.reason,
                )
                return None

            if result.is_modified:
                # MODIFIED - log and continue with new event
                logger.info(
                    "Event %s MODIFIED by %s: %s",
                    current_event.type,
                    interceptor.name,
                    result.reason,
                )

            # Continue with (possibly modified) event
            # result.event is guaranteed non-None here since not blocked
            assert result.event is not None
            current_event = result.event

        return current_event

    def _emit_to_handlers(self, event: Event) -> None:
        """Emit event to all subscribed handlers.

        Args:
            event: The event to emit.
        """
        handlers = self._subscribers.get(event.type, [])
        for handler in handlers:
            handler(event)

    def get_history(self) -> list[Event]:
        """Get a copy of all published events.

        Returns:
            List of events in chronological order (oldest first).
        """
        return list(self._history)

    def get_blocked_events(self) -> list[BlockedEvent]:
        """Get a copy of all blocked events.

        The blocked events audit channel records every event that was
        stopped by an interceptor, including the blocking reason.

        Returns:
            List of BlockedEvent records in chronological order.
        """
        # Import here to satisfy type checker when called
        from babylon.engine.interceptor import BlockedEvent  # noqa: F401

        return list(self._blocked_events)

    def clear_history(self) -> None:
        """Remove all events from history."""
        self._history.clear()

    def clear_blocked_events(self) -> None:
        """Remove all blocked event records."""
        self._blocked_events.clear()

    @property
    def interceptor_count(self) -> int:
        """Number of registered interceptors."""
        return len(self._interceptors)

File: rust/crates/babylon-kernel/src/event_bus.rs
//! Deterministic event bus: ports `kernel/event_bus.py`'s ordering
//! guarantees (spec §6.5), generic over the topic type — the 100-value
//! `EventType` domain enum lands with `babylon-domain`/`babylon-engine` in
//! Phase 2/3, not here.
//!
//! The four guarantees extracted from the Python source read end-to-end
//! (F1 discipline, 2026-07-30 — the fourth is one the Phase-1 plan's own
//! summary missed):
//!
//! 1. **Registration-order dispatch**: handlers for one topic fire in
//!    subscription order.
//! 2. **Append-before-emit**: the event enters the bus's history before
//!    any handler runs — so the event is never lost even when the fan-out
//!    fails.
//! 3. **Interceptor chain**: stable sort by priority DESCENDING (higher
//!    first), registration order as the tiebreak, sorted per publish.
//!    An interceptor may allow, BLOCK (the ORIGINAL event is recorded to a
//!    blocked-events audit channel with the interceptor's name and reason;
//!    nothing enters history, nothing dispatches), or MODIFY (the modified
//!    event continues down the chain, and it is the MODIFIED event that
//!    history records and handlers see).
//! 4. **Handler isolation** (`_emit_to_handlers`'s III.7 clause): every
//!    subscribed handler runs even when an earlier one fails; failures are
//!    collected and returned together AFTER the fan-out — isolation
//!    without silent swallowing. Rust surfaces Python's `ExceptionGroup`
//!    as an error list in `publish`'s `Result`.

/// One simulation event: a topic plus the tick it occurred on. The Python
/// `Event.payload` dict and its sim-time timestamp are domain-layer
/// concerns (Phase 2/3); the kernel bus pins the ordering machinery only.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Event<T> {
    /// The event's topic (the Python `Event.type` string).
    pub topic: T,
    /// The tick the event occurred on.
    pub tick: u64,
}

/// One handler failure, reported after the full fan-out (guarantee 4).
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct HandlerFailure {
    /// Index of the failing handler in registration order.
    pub handler_index: usize,
    /// The handler's own description of the failure.
    pub message: String,
}

/// What an interceptor decides for one event (the Python
/// `InterceptResult` surface).
#[derive(Debug, Clone)]
pub enum Intercept<T> {
    /// Pass the event through unchanged.
    Allow,
    /// Stop the event: it is recorded to the blocked-events audit channel
    /// and never reaches history or handlers.
    Block {
        /// Why the interceptor blocked it (audit-channel record).
        reason: String,
    },
    /// Replace the event; the replacement continues down the chain.
    Modify(Event<T>),
}

/// An interceptor: adversarial-mechanics middleware ahead of emission.
pub trait Interceptor<T> {
    /// Higher priority runs first; ties run in registration order.
    fn priority(&self) -> i32;
    /// The interceptor's audit-channel name.
    fn name(&self) -> &str;
    /// Decide this event's fate.
    fn intercept(&mut self, event: &Event<T>) -> Intercept<T>;
}

/// One blocked-event audit record: the ORIGINAL event (pre-modification —
/// the Python bus logs `event`, not `current_event`, for auditability),
/// plus who blocked it and why.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct BlockedEvent<T> {
    /// The original event as published.
    pub event: Event<T>,
    /// The blocking interceptor's name.
    pub interceptor_name: String,
    /// The blocking reason.
    pub reason: String,
}

/// A subscribed handler: returns `Err(message)` to report failure without
/// stopping the fan-out (guarantee 4).
pub type Handler<T> = Box<dyn FnMut(&Event<T>) -> Result<(), String>>;

/// The deterministic publish/subscribe bus.
pub struct EventBus<T> {
    subscribers: Vec<(T, Handler<T>)>,
    history: Vec<Event<T>>,
    interceptors: Vec<Box<dyn Interceptor<T>>>,
    blocked: Vec<BlockedEvent<T>>,
}

impl<T: Clone + PartialEq> EventBus<T> {
    /// An empty bus.
    #[must_use]
    pub fn new() -> Self {
        Self {
            subscribers: Vec::new(),
            history: Vec::new(),
            interceptors: Vec::new(),
            blocked: Vec::new(),
        }
    }

    /// Subscribe a handler to a topic. Registration order is dispatch
    /// order (guarantee 1).
    pub fn subscribe(&mut self, topic: T, handler: Handler<T>) {
        self.subscribers.push((topic, handler));
    }

    /// Register an interceptor. Sorted per publish — priority descending,
    /// registration order as the stable tiebreak (guarantee 3).
    pub fn register_interceptor(&mut self, interceptor: Box<dyn Interceptor<T>>) {
        self.interceptors.push(interceptor);
    }

    /// Number of registered interceptors.
    #[must_use]
    pub fn interceptor_count(&self) -> usize {
        self.interceptors.len()
    }

    /// Publish one event.
    ///
    /// Fast path with no interceptors (the Python zero-overhead clause);
    /// otherwise the chain runs first and a blocked event goes to the
    /// audit channel instead of history.
    ///
    /// # Errors
    /// Returns every [`HandlerFailure`] the fan-out produced, AFTER every
    /// handler has run (guarantee 4) — the event is already in history
    /// regardless (guarantee 2).
    pub fn publish(&mut self, event: Event<T>) -> Result<(), Vec<HandlerFailure>> {
        let event = if self.interceptors.is_empty() {
            event
        } else {
            match self.run_interceptors(event) {
                Some(processed) => processed,
                None => return Ok(()), // blocked: audited, nothing dispatched
            }
        };
        self.history.push(event.clone());
        self.dispatch(&event)
    }

    /// The interceptor chain: returns the surviving (possibly modified)
    /// event, or `None` if blocked (audit record written).
    fn run_interceptors(&mut self, event: Event<T>) -> Option<Event<T>> {
        // Stable sort of INDICES by descending priority: equal priorities
        // keep registration order, and sorting per publish mirrors the
        // Python bus exactly.
        let mut order: Vec<usize> = (0..self.interceptors.len()).collect();
        order.sort_by_key(|&i| std::cmp::Reverse(self.interceptors[i].priority()));

        let original = event.clone();
        let mut current = event;
        for i in order {
            let interceptor = &mut self.interceptors[i];
            match interceptor.intercept(&current) {
                Intercept::Allow => {}
                Intercept::Block { reason } => {
                    let name = interceptor.name().to_owned();
                    self.blocked.push(BlockedEvent {
                        event: original,
                        interceptor_name: name,
                        reason,
                    });
                    return None;
                }
                Intercept::Modify(replacement) => current = replacement,
            }
        }
        Some(current)
    }

    /// The isolated fan-out (guarantee 4): every matching handler runs;
    /// failures collect and return together.
    fn dispatch(&mut self, event: &Event<T>) -> Result<(), Vec<HandlerFailure>> {
        let mut failures = Vec::new();
        for (index, (topic, handler)) in self.subscribers.iter_mut().enumerate() {
            if *topic == event.topic {
                if let Err(message) = handler(event) {
                    failures.push(HandlerFailure {
                        handler_index: index,
                        message,
                    });
                }
            }
        }
        if failures.is_empty() {
            Ok(())
        } else {
            Err(failures)
        }
    }

    /// The published-event history, oldest first.
    #[must_use]
    pub fn history(&self) -> &[Event<T>] {
        &self.history
    }

    /// The blocked-events audit channel, oldest first.
    #[must_use]
    pub fn blocked_events(&self) -> &[BlockedEvent<T>] {
        &self.blocked
    }

    /// Clear the published-event history.
    pub fn clear_history(&mut self) {
        self.history.clear();
    }
}

impl<T: Clone + PartialEq> Default for EventBus<T> {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::{Event, EventBus, Intercept, Interceptor};
    use std::cell::RefCell;
    use std::rc::Rc;

    fn event(topic: &'static str, tick: u64) -> Event<&'static str> {
        Event { topic, tick }
    }

    #[test]
    fn handlers_fire_in_registration_order() {
        let mut bus: EventBus<&'static str> = EventBus::new();
        let order = Rc::new(RefCell::new(Vec::<u32>::new()));
        let o1 = Rc::clone(&order);
        let o2 = Rc::clone(&order);
        bus.subscribe(
            "t",
            Box::new(move |_| {
                o1.borrow_mut().push(1);
                Ok(())
            }),
        );
        bus.subscribe(
            "t",
            Box::new(move |_| {
                o2.borrow_mut().push(2);
                Ok(())
            }),
        );
        bus.publish(event("t", 0)).unwrap();
        assert_eq!(*order.borrow(), vec![1, 2]);
    }

    #[test]
    fn a_failing_handler_does_not_stop_later_handlers_and_all_failures_report() {
        // Guarantee 4 — the one the plan's summary missed. Python raises an
        // ExceptionGroup AFTER the fan-out; Rust returns the collected list.
        let mut bus: EventBus<&'static str> = EventBus::new();
        let ran = Rc::new(RefCell::new(Vec::<u32>::new()));
        let r1 = Rc::clone(&ran);
        let r3 = Rc::clone(&ran);
        bus.subscribe(
            "t",
            Box::new(move |_| {
                r1.borrow_mut().push(1);
                Ok(())
            }),
        );
        bus.subscribe("t", Box::new(|_| Err("second handler failed".into())));
        bus.subscribe(
            "t",
            Box::new(move |_| {
                r3.borrow_mut().push(3);
                Ok(())
            }),
        );
        let failures = bus.publish(event("t", 5)).unwrap_err();
        assert_eq!(*ran.borrow(), vec![1, 3], "later handlers must still run");
        assert_eq!(failures.len(), 1);
        assert_eq!(failures[0].handler_index, 1);
    }

    #[test]
    fn event_survives_in_history_even_when_every_handler_fails() {
        // Guarantee 2, in the form the Python docstring gives as its
        // REASON: append-before-emit means the event is never lost.
        let mut bus: EventBus<&'static str> = EventBus::new();
        bus.subscribe("t", Box::new(|_| Err("boom".into())));
        let _ = bus.publish(event("t", 9));
        assert_eq!(bus.history(), &[event("t", 9)]);
    }

    #[test]
    fn untargeted_topics_do_not_dispatch() {
        let mut bus: EventBus<&'static str> = EventBus::new();
        let ran = Rc::new(RefCell::new(0u32));
        let r = Rc::clone(&ran);
        bus.subscribe(
            "other",
            Box::new(move |_| {
                *r.borrow_mut() += 1;
                Ok(())
            }),
        );
        bus.publish(event("t", 0)).unwrap();
        assert_eq!(*ran.borrow(), 0);
        assert_eq!(
            bus.history().len(),
            1,
            "history records regardless of subscribers"
        );
    }

    // ---- interceptor chain ----

    struct Recorder {
        label: &'static str,
        priority: i32,
        seen: Rc<RefCell<Vec<&'static str>>>,
        verdict: fn(&Event<&'static str>) -> Intercept<&'static str>,
    }

    impl Interceptor<&'static str> for Recorder {
        fn priority(&self) -> i32 {
            self.priority
        }
        fn name(&self) -> &str {
            self.label
        }
        fn intercept(&mut self, event: &Event<&'static str>) -> Intercept<&'static str> {
            self.seen.borrow_mut().push(self.label);
            (self.verdict)(event)
        }
    }

    fn allow(_: &Event<&'static str>) -> Intercept<&'static str> {
        Intercept::Allow
    }

    #[test]
    fn interceptors_run_priority_desc_with_registration_tiebreak() {
        let mut bus: EventBus<&'static str> = EventBus::new();
        let seen = Rc::new(RefCell::new(Vec::new()));
        for (label, priority) in [("low", 1), ("high", 9), ("mid_a", 5), ("mid_b", 5)] {
            bus.register_interceptor(Box::new(Recorder {
                label,
                priority,
                seen: Rc::clone(&seen),
                verdict: allow,
            }));
        }
        bus.publish(event("t", 0)).unwrap();
        // Descending priority; mid_a before mid_b because registered first.
        assert_eq!(*seen.borrow(), vec!["high", "mid_a", "mid_b", "low"]);
    }

    #[test]
    fn a_blocked_event_is_audited_and_never_reaches_history_or_handlers() {
        let mut bus: EventBus<&'static str> = EventBus::new();
        let ran = Rc::new(RefCell::new(0u32));
        let r = Rc::clone(&ran);
        bus.subscribe(
            "t",
            Box::new(move |_| {
                *r.borrow_mut() += 1;
                Ok(())
            }),
        );
        bus.register_interceptor(Box::new(Recorder {
            label: "censor",
            priority: 0,
            seen: Rc::new(RefCell::new(Vec::new())),
            verdict: |_| Intercept::Block {
                reason: "state repression".into(),
            },
        }));
        bus.publish(event("t", 3)).unwrap();
        assert_eq!(*ran.borrow(), 0);
        assert!(bus.history().is_empty());
        let blocked = bus.blocked_events();
        assert_eq!(blocked.len(), 1);
        assert_eq!(blocked[0].interceptor_name, "censor");
        assert_eq!(blocked[0].event, event("t", 3));
    }

    #[test]
    fn a_modified_event_is_what_history_and_handlers_see_but_audit_keeps_the_original() {
        let mut bus: EventBus<&'static str> = EventBus::new();
        // High-priority modifier retargets tick 1 -> tick 100...
        bus.register_interceptor(Box::new(Recorder {
            label: "modifier",
            priority: 9,
            seen: Rc::new(RefCell::new(Vec::new())),
            verdict: |e| {
                Intercept::Modify(Event {
                    topic: e.topic,
                    tick: 100,
                })
            },
        }));
        bus.publish(event("t", 1)).unwrap();
        assert_eq!(bus.history(), &[event("t", 100)]);

        // ...and when a LOWER-priority blocker fires after a modifier, the
        // audit channel records the ORIGINAL event, per the Python bus.
        bus.register_interceptor(Box::new(Recorder {
            label: "blocker",
            priority: 0,
            seen: Rc::new(RefCell::new(Vec::new())),
            verdict: |_| Intercept::Block {
                reason: "after modify".into(),
            },
        }));
        bus.publish(event("t", 2)).unwrap();
        assert_eq!(
            bus.blocked_events()[0].event,
            event("t", 2),
            "original, not modified"
        );
    }
}
